pytest_rerunfailures.py

```diff
@@ -226,40 +226,45 @@ def pytest_runtest_protocol(item, nextitem):
     parallel = hasattr(item.config, "slaveinput")
     item.execution_count = 0
 
-    need_to_run = True
-    while need_to_run:
+    need_to_rerun = True
+    while need_to_rerun:
         item.execution_count += 1
         item.ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
         reports = runtestprotocol(item, nextitem=nextitem, log=False)
 
+        need_to_rerun = False
         for report in reports:  # 3 reports: setup, call, teardown
             is_terminal_error = _should_hard_fail_on_error(item.session.config, report)
+            out_of_runs = item.execution_count > reruns
             report.rerun = item.execution_count - 1
             xfail = hasattr(report, "wasxfail")
             if (
-                item.execution_count > reruns
+                out_of_runs
                 or not report.failed
                 or xfail
                 or is_terminal_error
+                or need_to_rerun
```

Comment:
With this way of writing we have 5 possible cases in this branch, which will never be properly addressed by any branch coverage. I would prefer a way, so that we need a testcase to make the coverage happy. And I would prefer a testcase.

Reply:
This hook was not called for reports of a test run beyond the first failure, which this PR fixes. So, before this PR if a test fails in phase `setup`, its `call` and `teardown` reports aren't processed ( in this case all we do for processing is call `pytest_runtest_logreport`

Now that I look at my change again I have a general question. Why do we not call `pytest_runtest_logreport` for failed test phases when xdist or parellel?